shell/browser/api/electron_api_web_contents.cc

```diff
@@ -2492,6 +2492,58 @@ content::NavigationEntry* WebContents::GetNavigationEntryAtIndex(
   return web_contents()->GetController().GetEntryAtIndex(index);
 }
 
+bool WebContents::DeleteNavigationEntryAtIndex(int index) {
+  if (index < 0 || index >= GetHistoryLength())
+    return false;
+
+  return web_contents()->GetController().RemoveEntryAtIndex(index);
+}
+
+std::vector<content::NavigationEntry*> WebContents::GetHistory() const {
+  const int history_length = GetHistoryLength();
+  auto& controller = web_contents()->GetController();
+
+  // If the history is empty, it contains only one entry and that is
+  // "InitialEntry"
+  if (history_length == 1 && controller.GetEntryAtIndex(0)->IsInitialEntry())
+    return std::vector<content::NavigationEntry*>();
+
+  std::vector<content::NavigationEntry*> history;
+  history.reserve(history_length);
+
+  for (int i = 0; i < history_length; i++)
+    history.push_back(controller.GetEntryAtIndex(i));
+
+  return history;
+}
+
+bool WebContents::ReplaceHistory(
+    const std::vector<gin_helper::Dictionary>& new_history,
+    int index) {
+  std::vector<std::unique_ptr<content::NavigationEntry>> history;
+  history.reserve(new_history.size());
+
+  for (const gin_helper::Dictionary& history_item : new_history) {
+    auto entry = content::NavigationEntry::Create();
+    entry->SetIsOverridingUserAgent(true);
+    std::u16string url, title;
+    if (history_item.Get("url", &url))
+      entry->SetURL(GURL(url));
+    if (history_item.Get("title", &title))
+      entry->SetTitle(title);
```

Comment:
My concern here is that ReplaceHistory may not correctly handle other important properties of the navigation entries, potentially leading to unexpected behavior or inconsistencies in the navigation history. In Electron's version of NavigationEntry, we are just exposing url and title. But Chromium's NavigationEntry has page state with scroll positions and other client side info/ referrer url and much more that might introduce inconsistencies if we just change the url and title. What do you think?

Reply:
@vitekcerny do you happen to have a reference to how Chrome restores navigation history? Maybe it'd provide us hints for how to make this more robust.

Reply:
Here is the implementation of the restore method: https://source.chromium.org/chromium/chromium/src/+/main:content/browser/renderer_host/navigation_controller_impl.cc;l=823
You can find all references by clicking on the name of the function.